lib/linked_list.rb

```diff
@@ -7,39 +7,65 @@ def initialize
     @head = nil
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(1)
+  # Space complexity - o(1)
   def add_first(data)
-
+    @head = Node.new(data, @head)
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(1)
+  # Space complexity - o(1)
   def get_first
-
+    return nil if head.nil?
+    return @head.data
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n)
+  # Space complexity - o(1)
   def length
-    return 0
+    return 0 if @head.nil?
+    current = @head
+    count = 0
+    while current != nil
+      count += 1
+      current = current.next
+    end
+    return count
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n)
+  # Space complexity - o(1)
   def add_last(data)
-
+    new_node = Node.new(data)
+    if @head.nil?
+      add_first(data)
+    else 
+      current = @head 
+      while current.next != nil
+       current = current.next
+      end
+      current.next = new_node
+    end
   end
 
-  # Time complexity - ?
-  # Space complexity - ?
+  # Time complexity - o(n)
+  # Space complexity - o(1)
   def get_last
-
+    current = @head
```

Comment:
What if `head` is `nil`?

```suggestion
    return nil if head.nil?
    current = head
```